ouroboros-network/test/Test/Ouroboros/Network/Testnet/Simulation/Node.hs

```diff
@@ -0,0 +1,346 @@
+                        -> RelayAccessPoint
+                        -> Gen [(Int, Map RelayAccessPoint PeerAdvertise)]
+      genLocalRootPeers l r = do
+        let newL = l \\ [r]
+            size = length newL
+        target <- chooseInt (1, size)
+        peerAdvertise <- vectorOf size arbitrary
+        let mapRelays = Map.fromList $ zip newL peerAdvertise
+
+        return [(target, mapRelays)]
+
+      -- | Given a NtNAddr generate the necessary things to run in Simulation
+      addressToRun :: [RelayAccessPoint]
+                   -> Map Domain [IP]
+                   -> (NtNAddr, RelayAccessPoint)
+                   -> Gen SimArgs
+      addressToRun raps dMap (ntnAddr, rap) = do
+        bgaSlotDuration <- fromInteger <$> choose (0, 100)
+        bgaSeed <- mkStdGen <$> arbitrary
+        quota <- chooseInt (0, 100)
+
+        -- These values approximately correspond to false positive
+        -- thresholds for streaks of empty slots with 99% probability,
+        -- 99.9% probability up to 99.999% probability.
+        -- t = T_s [log (1-Y) / log (1-f)]
+        -- Y = [0.99, 0.999...]
+        --
+        -- T_s = slot length of 1s.
+        -- f = 0.05
+        -- The timeout is randomly picked per bearer to avoid all bearers
+        -- going down at the same time in case of a long streak of empty
+        -- slots. TODO: workaround until peer selection governor.
+        -- Taken from ouroboros-consensus/src/Ouroboros/Consensus/Node.hs
+        mustReplyTimeout <- Just <$> oneof (pure <$> [90, 135, 180, 224, 269])
+
+        stdGen <- mkStdGen <$> arbitrary
+
+        lrp <- genLocalRootPeers raps rap
+
+        peerSelectionTargets <- arbitrary
+        dnsTimeout <- arbitrary
+        dnsLookupDelay <- arbitrary
+
+        return ( bgaSlotDuration
+               , bgaSeed
+               , quota
+               , mustReplyTimeout
+               , raps
+               , stdGen
+               , dMap
+               , ntnAddr
+               , lrp
+               , peerSelectionTargets
+               , dnsTimeout
+               , dnsLookupDelay
+               )
+
+-- | Run an arbitrary topology
+multinodeSim :: forall m. ( MonadAsync m
+                         , MonadFork m
+                         , MonadST m
+                         , MonadEvaluate m
+                         , MonadLabelledSTM m
+                         , MonadCatch       m
+                         , MonadMask        m
+                         , MonadTime        m
+                         , MonadTimer       m
+                         , MonadThrow  (STM m)
+                         , Eq (Async m Void)
+                         , forall a. Semigroup a => Semigroup (m a)
+                         )
+                      => BearerInfo
+                      -> MultiNodeScript
+                      -> m Void
+multinodeSim
+  defaultBearerInfo
+  (MultiNodeScript args) = withAsyncAll (map runNode args) $ \nodes -> do
```

Comment:
* It would be useful to vary when each node joins the system.
* It would be nice to verify what happens if a node is shutdown with an async exception, will it be able to rejoin the network again?

This suggests we actually need a script with a few instructions:
* `JoinNetwork`
* `Kill`
anything else that you think would be useful?

Reply:
> It would be useful to vary when each node joins the system.

You mean a delay ?

> It would be nice to verify what happens if a node is shutdown with an async exception, will it be able to rejoin the network again?

I agree I just decided to come up with simplest setting possible.

> This suggests we actually need a script with a few instructions:
>
>  -  JoinNetwork
>  -  Kill
>
>    anything else that you think would be useful?

I don't think there's much more that can happen on a real scenario, right?

Reply:
Yes a startup delay.  It seems we can only  came up with the two :).  But one could be able to chain them `[JoinNetwork, Kill, ... ]`.  The generators written by Ulf are a good example hot to generate stateful instructions.  Also note that we don't need many of them, a list of 2 / 3 on average, 5-8 at max should be ok.  At the first iteration we rather want to explore the case you were thinking about: all nodes are up and running, and maybe some times some of them go down and re-join.  But later this will allow us to extend the generator to include situations like: all nodes going down at the same time, and stay down for some time, rejoin simultaneously.

There's another aspect of this, and needs explanation.  We can simulate network errors through attenuation.  I argue that network attenuation never bring a node down, so there's no overlap, making this quite a bit simpler.  Well, unless we inject an `accept` error some of them can bring the node down, I think that's the only case (do you agree?).  If so we could ignore this (but it should be described, that the tests are designed with that in mind).   The problem with this is that we'd have two ways of bringing the node down: `Kill` and an accept exception.  This means we cannot relay on the interval between a node being killed and re-joining the network, as it could be in that state for longer.  This seems like a minor caveat, since probably none of ours properties will relay on this information (am I right?).

Reply:
So if I understood correctly, we want that each node have a script with instructions + some delay (in the same way `Server2` `MultiNodeScript` does). 

And regarding your point in the second paragraph, you are saying to ignore `accept` errors since we do not/won't inject any? And that we can not write tests relying on just `Kill` command alone because it could have been killed for longer due to a possible `accept` exception?

Reply:
> So if I understood correctly, we want that each node have a script with instructions + some delay (in the same way Server2 MultiNodeScript does).
Yes, that's right.

> And regarding your point in the second paragraph, you are saying to ignore accept errors since we do not/won't inject any? And that we can not write tests relying on just Kill command alone because it could have been killed for longer due to a possible accept exception?

My original thought was that we at least we don't inject errors that would kill the node, because this makes it harder to make a script which `Kill` will actually kill a node.  But, giving it another thought, I think we don't need to worry about it at all, `Kill` if a  nodes does not work would be no-op, and it would not affect `Start` (for which we need to track if the node is not running, but if it stopped earlier due to some exception, that's ok).

One additional useful command would be `Reconfigure` which reconfigures the local / public root peers.

Reply:
Addressed